src/realm/object-store/sync/app_user.hpp

```diff
@@ -0,0 +1,267 @@
+    {
+        return get_field("first_name");
+    }
+    // The last name of the user.
+    std::optional<std::string> last_name() const
+    {
+        return get_field("last_name");
+    }
+    // The gender of the user.
+    std::optional<std::string> gender() const
+    {
+        return get_field("gender");
+    }
+    // The birthdate of the user.
+    std::optional<std::string> birthday() const
+    {
+        return get_field("birthday");
+    }
+    // The minimum age of the user.
+    std::optional<std::string> min_age() const
+    {
+        return get_field("min_age");
+    }
+    // The maximum age of the user.
+    std::optional<std::string> max_age() const
+    {
+        return get_field("max_age");
+    }
+
+    bson::Bson operator[](const std::string& key) const
+    {
+        return m_data.at(key);
+    }
+
+    const bson::BsonDocument& data() const
+    {
+        return m_data;
+    }
+
+    UserProfile(bson::BsonDocument&& data)
+        : m_data(std::move(data))
+    {
+    }
+    UserProfile() = default;
+
+private:
+    bson::BsonDocument m_data;
+
+    std::optional<std::string> get_field(const char* name) const
+    {
+        if (auto val = m_data.find(name)) {
+            return static_cast<std::string>((*val));
+        }
+        return util::none;
+    }
+};
+
+// A struct that represents an identity that a `User` is linked to
+struct UserIdentity {
+    // the id of the identity
+    std::string id;
+    // the associated provider type of the identity
+    std::string provider_type;
+
+    UserIdentity(const std::string& id, const std::string& provider_type);
+
+    bool operator==(const UserIdentity& other) const
+    {
+        return id == other.id && provider_type == other.provider_type;
+    }
+
+    bool operator!=(const UserIdentity& other) const
+    {
+        return id != other.id || provider_type != other.provider_type;
+    }
+};
+
+struct UserData {
+    // Current refresh token or empty if user is logged out
+    RealmJWT refresh_token;
+    // Current access token or empty if user is logged out
+    RealmJWT access_token;
+    // UUIDs which used to be used to generate local Realm file paths. Now only
+    // used to locate existing files.
+    std::vector<std::string> legacy_identities;
+    // Identities which were used to log into this user
+    std::vector<UserIdentity> identities;
+    // Id for the device which this user was logged in on. Users are not
+    // portable between devices so this cannot be changed after the user
+    // is created
+    std::string device_id;
+    // Server-stored user profile
+    UserProfile profile;
+};
+
+class User final : public SyncUser, public std::enable_shared_from_this<User>, public Subscribable<User> {
+    struct Private {};
+
+public:
+    // ------------------------------------------------------------------------
+    // SyncUser implementation
+
+    std::string user_id() const noexcept override;
```

Comment:
Why not simply `id` ... its already namespaced by `User`?

Reply:
Most of the places where user_id is used isn't in User.